File: Problems/python/boj_바닥장식.py
#https://www.acmicpc.net/problem/1388
#DFS

import sys
input = sys.stdin.readline

n, m = map(int, input().split())
mat = [list(input().rstrip()) for _ in range(n)]

def dfs(d,x,y):
    mat[x][y]=0
    if d=='-':
        if not 0<=y+1<m:
            return
        if d==mat[x][y+1]:
            dfs(d,x,y+1)
    else:
        if not 0<=x+1<n:
            return
        if d==mat[x+1][y]:
            dfs(d,x+1,y)

count = 0    
for i in range(n):
    for j in range(m):
        if mat[i][j]!=0:
            dfs(mat[i][j],i,j)
            count += 1

print(count)

File: Problems/cpp/boj_바닥장식.cpp
//https://www.acmicpc.net/problem/1388
//DFS, Implementation

#include <iostream>
#include <vector>
#include <string>

using namespace std;

vector<string> mat;
int n, m, result=0;

void dfs(char a, int i, int j){
    if (a=='-'){
        j++;
        if (j>=m) return;
        else if (mat[i][j]==a){
            mat[i][j] = 0;
            dfs(a, i, j);
        }
    }
    else{
        i++;
        if (i>=n) return;
        else if (mat[i][j]==a){
            mat[i][j] = 0;
            dfs(a, i, j);
        }
    }
}

int main()
{
	ios::sync_with_stdio(0);
	cin.tie(0);

	cin >> n >> m;

    string tmp;
	for (int i = 0; i < n; i++){
        cin >> tmp;
        mat.push_back(tmp);
    }

    for (int i=0; i<n; i++){
        for (int j=0; j<m; j++){
            if (mat[i][j]!=0){
                dfs(mat[i][j], i, j);
                result++;
            }
        }
    }

	cout << result;
}